packages/block-editor/src/components/block-popover/index.js

```diff
@@ -88,6 +124,49 @@ function BlockPopover(
 
 		return {
 			getBoundingClientRect() {
+				let postRootElement;
+				// The zoom out view has a vertical block toolbar that should always
+				// be on the edge of the canvas, aligned to the top of the currently
+				// selected section. This condition changes the anchor of the toolbar
+				// to the section instead of the block to handle blocks that are
+				// not full width and nested blocks to keep section height.
+				if ( isZoomOut && isSectionSelected ) {
+					// If the rootSectionElement is undefined then recurse up the DOM tree
+					// to find the element with `block-editor-iframe__body` classname.
+					// This can then be used as the anchor point.
+					// FIXME: we should not rely on classnames to find the
+					// root section element
+					if ( ! rootSectionElement ) {
+						postRootElement = selectedElement.closest(
+							'.block-editor-iframe__body'
+						);
+					}
+
+					// Compute the height based on the parent section of the
+					// selected block, because the selected block may be
+					// shorter than the section.
+					const rootSectionElementRect = getVisibleElementBounds(
+						rootSectionElement || postRootElement
+					);
+					const parentSectionElementRect =
+						getVisibleElementBounds( parentSectionElement );
+					const anchorHeight =
+						parentSectionElementRect.bottom -
+						parentSectionElementRect.top;
+
+					// Always use the width of the section root element to make sure
+					// the toolbar is always on the edge of the canvas.
+					const andchorWidth =
+						rootSectionElementRect.right -
+						rootSectionElementRect.left;
+					return new window.DOMRectReadOnly(
+						rootSectionElementRect.left,
```

Comment:
Disclaimer - this PR provides a better experience than trunk, so it's a good improvement, and I don't think this comment is a blocker to merge, the positioning is something that can be iteratively improved ...

----

For the horizontal positioning the section root might not be full width, which makes this unreliable. For example, the `<main>` element can be wrapped in another group that adds side padding or created a non-full-width layout. Here's an example where I created that situation while I had this PR checked out:
![Uploading Screenshot 2024-10-14 at 3.22.49 pm.png…]()

I think a safer alternative could be an API that returns a ref to the canvas element or the rect of the canvas.

Or it could be a slot next to the canvas that the toolbar renders in, and then the horizontal positioning wouldn't matter, only the vertical positioning.

Reply:
Hmm, both suggestions are smart. Let's see if thre is any simple approach to either of them.

Reply:
Here’s some steps to reproduce (I tested using Twenty Twenty Five):
1. In the site editor, open the Pages template
2. Find the Group block that acts as <main>  (it should be the one at root level)
3. Select it and from the block settings menu Group it to wrap it in a group (after this, the <main> group should become content width and not full width)
4. Save the template
5. Still in the site editor, open the Pages and edit a page (e.g. ‘Sample Page’)
6. Enter Zoomed Out mode and select a section
7. Observe that the vertical toolbar is on the canvas

Reply:
We already in this PR get the `.block-editor-iframe__body` as a fallback but we assumed that the main group should always be 1st choice. Maybe it shouldn't?